File: npm/marquette/src/validate.ts
import type {
  ApplicationMarquette,
  MARQUETTE_FORMAT_VERSION,
  MarquetteEnvironment,
  MarquetteRoute,
  MarquetteTarget,
  RenderingMode,
} from "./model.js";

/** Severity of an application-marquette diagnostic. */
export type MarquetteDiagnosticSeverity = "error" | "warning";

/** Stable, source-addressable application-marquette diagnostic. */
export interface MarquetteDiagnostic {
  /** Stable machine-readable diagnostic code. */
  readonly code: `VIZE_MARQUETTE_${string}`;
  /** Severity used by CLI, editor, test, and CI consumers. */
  readonly severity: MarquetteDiagnosticSeverity;
  /** JSON-style path into the authored marquette. */
  readonly path: string;
  /** Human-readable explanation and next action. */
  readonly message: string;
}

const IDENTIFIER = /^[a-z0-9][a-z0-9._-]*$/;
const SUPPORTED_FORMAT_VERSION: typeof MARQUETTE_FORMAT_VERSION = 1;

/**
 * Validates a complete application marquette.
 *
 * Diagnostics cover structural, reference, capability, target, and dependency
 * invariants. Results are sorted by path, code, and message so identical input
 * produces stable editor, CLI, test, and CI output.
 *
 * This function does not mutate the authored object and does not throw.
 */
export function validateApplicationMarquette(
  marquette: ApplicationMarquette,
): MarquetteDiagnostic[] {
  const diagnostics: MarquetteDiagnostic[] = [];
  const environments = marquette.environments ?? [];
  const backends = marquette.backends ?? [];
  const protocols = marquette.protocols ?? [];
  const routes = marquette.routes ?? [];
  const targets = new Set(marquette.targets ?? []);
  const capabilityIds = new Set(Object.keys(marquette.capabilities ?? {}));

  if ((marquette.formatVersion ?? SUPPORTED_FORMAT_VERSION) !== SUPPORTED_FORMAT_VERSION) {
    diagnostics.push(
      error("VIZE_MARQUETTE_001", "formatVersion", "unsupported marquette format version"),
    );
  }

  validateIdentifier(marquette.application, "application", "VIZE_MARQUETTE_002", diagnostics);

  for (const [key, capability] of Object.entries(marquette.capabilities ?? {})) {
    const path = contractPath("capabilities", key);
    validateIdentifier(key, path, "VIZE_MARQUETTE_003", diagnostics);
    if (key !== capability.id) {
      diagnostics.push(
        error("VIZE_MARQUETTE_004", path, "capability map key must equal capability.id"),
      );
    }
    const version = capability.version ?? 1;
    if (!Number.isInteger(version) || version < 1) {
      diagnostics.push(
        error("VIZE_MARQUETTE_005", path, "capability version must be greater than zero"),
      );
    }
    if (capability.description.trim().length === 0) {
      diagnostics.push(
        error("VIZE_MARQUETTE_024", path, "capability description must not be empty"),
      );
    }
  }

  const environmentIds = collectUniqueIds("environments", environments, diagnostics);
  const backendIds = collectUniqueIds("backends", backends, diagnostics);
  const protocolIds = collectUniqueIds("protocols", protocols, diagnostics);
  collectUniqueIds("routes", routes, diagnostics);

  for (const environment of environments) {
    const path = contractPath("environments", environment.id);
    if (!targets.has(environment.target)) {
      diagnostics.push(
        error("VIZE_MARQUETTE_007", path, "environment target must be declared in targets"),
      );
    }
    for (const dependency of [...new Set(environment.dependsOn ?? [])].sort()) {
      if (dependency === environment.id) {
        diagnostics.push(error("VIZE_MARQUETTE_008", path, "environment cannot depend on itself"));
      } else if (!environmentIds.has(dependency)) {
        diagnostics.push(
          error("VIZE_MARQUETTE_009", path, "environment dependency does not exist"),
        );
      }
    }
    validateCapabilities(path, environment.capabilities, capabilityIds, diagnostics);

    if (
      environment.consumer === "client" &&
      (environment.runtime === "rust" ||
        environment.runtime === "go" ||
        environment.runtime === "jvm")
    ) {
      diagnostics.push(
        warning(
          "VIZE_MARQUETTE_010",
          path,
          "client environment uses a server-oriented runtime; declare an adapter capability if this is intentional",
        ),
      );
    }
  }

  validateEnvironmentCycles(environments, diagnostics);

  for (const backend of backends) {
    const path = contractPath("backends", backend.id);
    if (backend.environment != null) {
      const environment = environments.find((candidate) => candidate.id === backend.environment);
      if (environment == null) {
        diagnostics.push(error("VIZE_MARQUETTE_011", path, "backend environment does not exist"));
      } else if (environment.consumer !== "server") {
        diagnostics.push(
          error("VIZE_MARQUETTE_012", path, "backend environment must be a server consumer"),
        );
      }
    }
    validateCapabilities(path, backend.capabilities, capabilityIds, diagnostics);
  }

  for (const protocol of protocols) {
    const path = contractPath("protocols", protocol.id);
    if (!backendIds.has(protocol.backend)) {
      diagnostics.push(error("VIZE_MARQUETTE_013", path, "protocol backend does not exist"));
    }
    validateCapabilities(path, protocol.capabilities, capabilityIds, diagnostics);
  }

  const routePaths = new Map<string, string>();
  for (const route of routes) {
    const path = contractPath("routes", route.id);
    if (!route.path.startsWith("/")) {
      diagnostics.push(error("VIZE_MARQUETTE_014", path, "route path must start with /"));
    }
    const environment = environments.find((candidate) => candidate.id === route.environment);
    if (environment == null) {
      diagnostics.push(error("VIZE_MARQUETTE_015", path, "route environment does not exist"));
    }
    if (route.backend != null && !backendIds.has(route.backend)) {
      diagnostics.push(error("VIZE_MARQUETTE_016", path, "route backend does not exist"));
    }
    if (route.protocol != null) {
      const protocol = protocols.find((candidate) => candidate.id === route.protocol);
      if (!protocolIds.has(route.protocol) || protocol == null) {
        diagnostics.push(error("VIZE_MARQUETTE_017", path, "route protocol does not exist"));
      } else if (route.backend != null && protocol.backend !== route.backend) {
        diagnostics.push(
          error(
            "VIZE_MARQUETTE_018",
            path,
            "route protocol and backend must refer to the same service",
          ),
        );
      }
    }
    if (environment != null && !renderingMatchesTarget(route, environment.target)) {
      diagnostics.push(
        error(
          "VIZE_MARQUETTE_023",
          path,
          "rendering mode is not compatible with the route environment target",
        ),
      );
    }
    validateCapabilities(path, route.capabilities, capabilityIds, diagnostics);

    const routeKey = `${route.environment}\u0000${route.path}`;
    const previous = routePaths.get(routeKey);
    if (previous != null) {
      diagnostics.push(
        error("VIZE_MARQUETTE_019", path, `route path is already used by route "${previous}"`),
      );
    }
    routePaths.set(routeKey, route.id);
  }

  for (const target of targets) {
    if (!environments.some((environment) => environment.target === target)) {
      diagnostics.push(
        warning("VIZE_MARQUETTE_020", "targets", "declared target has no environment"),
      );
    }
  }

  return diagnostics.sort(compareDiagnostics);
}

function collectUniqueIds(
  collection: string,
  values: readonly { readonly id: string }[],
  diagnostics: MarquetteDiagnostic[],
): Set<string> {
  const ids = new Set<string>();
  for (const value of values) {
    const path = contractPath(collection, value.id);
    validateIdentifier(value.id, path, "VIZE_MARQUETTE_006", diagnostics);
    if (ids.has(value.id)) {
      diagnostics.push(
        error("VIZE_MARQUETTE_006", path, "identifier must be unique within its collection"),
      );
    }
    ids.add(value.id);
  }
  return ids;
}

function validateIdentifier(
  id: string,
  path: string,
  code: `VIZE_MARQUETTE_${string}`,
  diagnostics: MarquetteDiagnostic[],
): void {
  if (!IDENTIFIER.test(id)) {
    diagnostics.push(
      error(
        code,
        path,
        "identifier must use lowercase ASCII letters, digits, dash, underscore, or dot",
      ),
    );
  }
}

function validateCapabilities(
  path: string,
  capabilities: readonly string[] | undefined,
  declared: ReadonlySet<string>,
  diagnostics: MarquetteDiagnostic[],
): void {
  for (const capability of [...new Set(capabilities ?? [])].sort()) {
    if (!declared.has(capability)) {
      diagnostics.push(error("VIZE_MARQUETTE_021", path, "referenced capability is not declared"));
    }
  }
}

function validateEnvironmentCycles(
  environments: readonly MarquetteEnvironment[],
  diagnostics: MarquetteDiagnostic[],
): void {
  const graph = new Map(environments.map((value) => [value.id, value.dependsOn ?? []]));
  const visiting = new Set<string>();
  const visited = new Set<string>();

  for (const id of [...graph.keys()].sort()) {
    if (hasCycle(id, graph, visiting, visited)) {
      diagnostics.push(
        error(
          "VIZE_MARQUETTE_022",
          contractPath("environments", id),
          "environment dependency graph must be acyclic",
        ),
      );
    }
  }
}

function hasCycle(
  id: string,
  graph: ReadonlyMap<string, readonly string[]>,
  visiting: Set<string>,
  visited: Set<string>,
): boolean {
  if (visited.has(id)) return false;
  if (visiting.has(id)) return true;
  visiting.add(id);
  const cyclic = [...(graph.get(id) ?? [])]
    .sort()
    .some((dependency) => graph.has(dependency) && hasCycle(dependency, graph, visiting, visited));
  visiting.delete(id);
  visited.add(id);
  return cyclic;
}

function renderingMatchesTarget(route: MarquetteRoute, target: MarquetteTarget): boolean {
  const targetRendering: Partial<Record<RenderingMode, MarquetteTarget>> = {
    native: "native",
    desktop: "desktop",
    terminal: "terminal",
  };
  return (targetRendering[route.rendering] ?? "web") === target;
}

function compareDiagnostics(left: MarquetteDiagnostic, right: MarquetteDiagnostic): number {
  return (
    compareText(left.path, right.path) ||
    compareText(left.code, right.code) ||
    compareText(left.message, right.message)
  );
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function contractPath(collection: string, id: string): string {
  return `${collection}.${id}`;
}

function error(
  code: `VIZE_MARQUETTE_${string}`,
  path: string,
  message: string,
): MarquetteDiagnostic {
  return { code, severity: "error", path, message };
}

function warning(
  code: `VIZE_MARQUETTE_${string}`,
  path: string,
  message: string,
): MarquetteDiagnostic {
  return { code, severity: "warning", path, message };
}

File: crates/vize_marquette/src/validate.rs
use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};
use vize_s0::{String, cstr};

use crate::{ApplicationContract, CONTRACT_FORMAT_VERSION, EnvironmentConsumer, RuntimeFamily};

pub(crate) mod rules;
#[cfg(test)]
mod tests;

use rules::{
    collect_unique_ids, contract_path, validate_capabilities, validate_environment_cycles,
    validate_identifier, validate_rendering_target,
};

/// Severity of an application-marquette diagnostic.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DiagnosticSeverity {
    /// Contract execution must stop.
    Error,
    /// Contract is valid but contains a likely configuration mistake.
    Warning,
}

/// Stable, source-addressable application-contract diagnostic.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ContractDiagnostic {
    /// Stable machine-readable diagnostic code.
    pub code: &'static str,
    /// Severity used by CLI, editor, and CI consumers.
    pub severity: DiagnosticSeverity,
    /// JSON-style path into the authored contract.
    pub path: String,
    /// Human-readable explanation and next action.
    pub message: String,
}

impl ContractDiagnostic {
    /// Creates an error diagnostic that prevents adapter execution.
    pub(crate) fn error(
        code: &'static str,
        path: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self {
            code,
            severity: DiagnosticSeverity::Error,
            path: path.into(),
            message: message.into(),
        }
    }

    /// Creates a warning diagnostic for a valid but suspicious contract.
    pub(crate) fn warning(
        code: &'static str,
        path: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self {
            code,
            severity: DiagnosticSeverity::Warning,
            path: path.into(),
            message: message.into(),
        }
    }
}

/// Validates a complete application contract.
///
/// Diagnostics are deterministic and sorted by path, code, and message so the
/// same contract produces stable editor, CLI, test, and CI output.
pub fn validate_contract(contract: &ApplicationContract) -> Vec<ContractDiagnostic> {
    let mut diagnostics = Vec::new();

    if contract.format_version != CONTRACT_FORMAT_VERSION {
        diagnostics.push(ContractDiagnostic::error(
            "VIZE_MARQUETTE_001",
            "formatVersion",
            "unsupported marquette format version",
        ));
    }

    validate_identifier(
        &contract.application,
        "application",
        "VIZE_MARQUETTE_002",
        &mut diagnostics,
    );

    let capability_ids = contract
        .capabilities
        .keys()
        .cloned()
        .collect::<BTreeSet<_>>();
    for (key, capability) in &contract.capabilities {
        validate_identifier(
            key,
            &contract_path("capabilities", key),
            "VIZE_MARQUETTE_003",
            &mut diagnostics,
        );
        if key.as_str() != capability.id.as_str() {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_004",
                contract_path("capabilities", key),
                "capability map key must equal capability.id",
            ));
        }
        if capability.version == 0 {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_005",
                contract_path("capabilities", key),
                "capability version must be greater than zero",
            ));
        }
        if capability.description.trim().is_empty() {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_024",
                contract_path("capabilities", key),
                "capability description must not be empty",
            ));
        }
    }

    let environment_ids = collect_unique_ids(
        "environments",
        contract.environments.iter().map(|value| &value.id),
        &mut diagnostics,
    );
    let backend_ids = collect_unique_ids(
        "backends",
        contract.backends.iter().map(|value| &value.id),
        &mut diagnostics,
    );
    let protocol_ids = collect_unique_ids(
        "protocols",
        contract.protocols.iter().map(|value| &value.id),
        &mut diagnostics,
    );
    collect_unique_ids(
        "routes",
        contract.routes.iter().map(|value| &value.id),
        &mut diagnostics,
    );

    for environment in &contract.environments {
        let path = contract_path("environments", &environment.id);
        if !contract.targets.contains(&environment.target) {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_007",
                path.clone(),
                "environment target must be declared in targets",
            ));
        }
        for dependency in &environment.depends_on {
            if dependency.as_str() == environment.id.as_str() {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_008",
                    path.clone(),
                    "environment cannot depend on itself",
                ));
            } else if !environment_ids.contains(dependency) {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_009",
                    path.clone(),
                    "environment dependency does not exist",
                ));
            }
        }
        validate_capabilities(
            &path,
            &environment.capabilities,
            &capability_ids,
            &mut diagnostics,
        );

        if environment.consumer == EnvironmentConsumer::Client
            && matches!(
                environment.runtime,
                RuntimeFamily::Rust | RuntimeFamily::Go | RuntimeFamily::Jvm
            )
        {
            diagnostics.push(ContractDiagnostic::warning(
                "VIZE_MARQUETTE_010",
                path,
                "client environment uses a server-oriented runtime; declare an adapter capability if this is intentional",
            ));
        }
    }

    validate_environment_cycles(contract, &mut diagnostics);

    for backend in &contract.backends {
        let path = contract_path("backends", &backend.id);
        if let Some(environment) = &backend.environment {
            if !environment_ids.contains(environment) {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_011",
                    path.clone(),
                    "backend environment does not exist",
                ));
            } else if contract
                .environments
                .iter()
                .find(|candidate| candidate.id.as_str() == environment.as_str())
                .is_some_and(|candidate| candidate.consumer != EnvironmentConsumer::Server)
            {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_012",
                    path.clone(),
                    "backend environment must be a server consumer",
                ));
            }
        }
        validate_capabilities(
            &path,
            &backend.capabilities,
            &capability_ids,
            &mut diagnostics,
        );
    }

    for protocol in &contract.protocols {
        let path = contract_path("protocols", &protocol.id);
        if !backend_ids.contains(&protocol.backend) {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_013",
                path.clone(),
                "protocol backend does not exist",
            ));
        }
        validate_capabilities(
            &path,
            &protocol.capabilities,
            &capability_ids,
            &mut diagnostics,
        );
    }

    let mut route_paths = BTreeMap::<(&String, &String), &String>::new();
    for route in &contract.routes {
        let path = contract_path("routes", &route.id);
        if !route.path.starts_with('/') {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_014",
                path.clone(),
                "route path must start with /",
            ));
        }
        if !environment_ids.contains(&route.environment) {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_015",
                path.clone(),
                "route environment does not exist",
            ));
        }
        if let Some(backend) = &route.backend
            && !backend_ids.contains(backend)
        {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_016",
                path.clone(),
                "route backend does not exist",
            ));
        }
        if let Some(protocol) = &route.protocol {
            if !protocol_ids.contains(protocol) {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_017",
                    path.clone(),
                    "route protocol does not exist",
                ));
            } else if let Some(route_backend) = &route.backend
                && contract
                    .protocols
                    .iter()
                    .find(|candidate| candidate.id.as_str() == protocol.as_str())
                    .is_some_and(|candidate| candidate.backend.as_str() != route_backend.as_str())
            {
                diagnostics.push(ContractDiagnostic::error(
                    "VIZE_MARQUETTE_018",
                    path.clone(),
                    "route protocol and backend must refer to the same service",
                ));
            }
        }
        validate_rendering_target(contract, route, &path, &mut diagnostics);
        validate_capabilities(
            &path,
            &route.capabilities,
            &capability_ids,
            &mut diagnostics,
        );

        let key = (&route.environment, &route.path);
        if let Some(previous) = route_paths.insert(key, &route.id) {
            diagnostics.push(ContractDiagnostic::error(
                "VIZE_MARQUETTE_019",
                path,
                cstr!("route path is already used by route \"{previous}\""),
            ));
        }
    }

    for target in &contract.targets {
        if !contract
            .environments
            .iter()
            .any(|environment| &environment.target == target)
        {
            diagnostics.push(ContractDiagnostic::warning(
                "VIZE_MARQUETTE_020",
                "targets",
                "declared target has no environment",
            ));
        }
    }

    diagnostics.sort_by(|left, right| {
        (&left.path, left.code, &left.message).cmp(&(&right.path, right.code, &right.message))
    });
    diagnostics
}
